### PR TITLE
parse import assertions

### PR DESCRIPTION
**Pull request by xtuc**
Refs https://github.com/webpack/webpack/issues/11917

**What kind of change does this PR introduce?**

Parsing import assertions syntax

**Did you add tests for your changes?**
Yes, one. Most of the tests are in https://github.com/xtuc/acorn-import-assertions/tree/main/test/fixtures

**Does this PR introduce a breaking change?**
no

**What needs to be documented once your changes are merged?**
probably not now, import assertions doesn't emit runtime code yet

### CODE REVIEW

**Comment by alexander-akait** [comment]
Tests are broken :disappointed:

**Comment by xtuc** [comment]
@alexander-akait yes noticed that too. I was wondering if those failures are related to my change? I can't see the detail of the failures but they mismatching stats? I'll try to run them locally again

**Comment by xtuc** [comment]
@alexander-akait tests are passing, coverage is failing and https://dev.azure.com/webpack/webpack/_build/results?buildId=12330&view=logs&j=1d505e51-909d-5717-0d5b-43f6a0bd49f6&t=6309f675-4460-5c1b-70bc-fd472dead250 is flaky. Could you please restart for me?

**Comment by alexander-akait** [comment]
@vankop thanks)

**Comment by xtuc** [comment]
right, sorry I forgot about that trick! thanks

**Comment by xtuc** [comment]
@sokra  could you please take a look at the PR? the idea is that we allow parsing and stripping out import assertions for now.
Next steps would be to emit runtime code to check the type of dynamically loaded resources.

**Comment by sokra** [comment]
> Next steps would be to emit runtime code to check the type of dynamically loaded resources.

JSON is not loaded dynamically but embedded into the JS file. As it's parsed during build it's already verified that it's really JSON and not JS or something else.

I think what we can do it to set the module type for imports with `assert { type: "json" }` to JSON so they go through the JSON parser. This could be done by allowing an `assert` condition in `module.rules` and adding a default rule to set `type: "json"` for `assert: { type: "json" }` (`module.defaultRules: [ ..., { assert: { type: "json" }, type: "json" }]` in `lib/config/defaults.js`).

For a `assert` condition we need to pass the `assert` value from `HarmonyImportDependencyParserPlugin` -> `ModuleDependency.assert` (`getResourceIdentifier` need to have `assert` too) -> `NormalModuleFactory` (add `assert` to `ruleSetCompiler`, and pass the value when executing it, get it from `dependencies[0].assert`).

**Comment by xtuc** [comment]
> JSON is not loaded dynamically but embedded into the JS file. As it's parsed during build it's already verified that it's really JSON and not JS or something else.

I was thinking of loading dynamically remote resources. Like: `import("https://a.com/great-json.php")`, the runtime should ensure that the response is indeed JSON. But we can do that later.

Thanks for your pointers, i'll do that change.

**Comment by xtuc** [comment]
@sokra  I believe I have implemented what you suggested. One test doesn't pass yet and we probably want to show a code frame when the assertion aren't met, that's still WIP. Could you please do early feedback on the implementation?

File: package.json [hunk 1]
```diff
@@ -11,6 +11,7 @@
     "@webassemblyjs/wasm-edit": "1.11.0",
     "@webassemblyjs/wasm-parser": "1.11.0",
     "acorn": "^8.0.4",
+    "acorn-import-assertions": "1.4.6",
```
Comment (alexander-akait) [inline package.json:14]: Can we use `^1.4.6`?

**Comment by xtuc** [comment]
@sokra could you please review the change so far?  I believe it matches what you suggested.

**Comment by sokra** [comment]
@xtuc `export { a } from "something" assert { type: "json" }` is missing in the parser

**Review by sokra — APPROVED** [review]

**Comment by sokra** [comment]
Thanks

**Comment by xtuc** [comment]
Great, thanks for your help!

**Comment by redonkulus** [comment]
@sokra is https://github.com/webpack/webpack/issues/13814 related to this merge?

**Comment by alexander-akait** [comment]
@redonkulus yep

**Comment by sodatea** [comment]
Should `assert` be mentioned in https://webpack.js.org/configuration/module/#rule ?
I've seen it used in `css-loader` examples but can't find it in the webpack documentation.

**Comment by alexander-akait** [comment]
@sodatea Yes, feel free to open an issue

**Pull request merged**